src/operator/reduce.ts

```diff
@@ -53,16 +53,24 @@ export function reduce<T>(this: Observable<T>, accumulator: (acc: T[], value: T,
 export function reduce<T, R>(this: Observable<T>, accumulator: (acc: R, value: T, index: number) => R, seed?: R): Observable<R>;
 /* tslint:disable:max-line-length */
 export function reduce<T, R>(this: Observable<T>, accumulator: (acc: R, value: T) => R, seed?: R): Observable<R> {
-  return this.lift(new ReduceOperator(accumulator, seed));
+  let hasSeed = false;
+  // providing a seed of `undefined` *should* be valid and trigger
+  // hasSeed! so don't use `seed !== undefined` checks!
+  // For this reason, we have to check it here at the original call site
+  // otherwise inside Operator/Subscriber we won't know if `undefined`
+  // means they didn't provide anything or if they literally provided `undefined`
+  if (arguments.length >= 2) {
+    hasSeed = true;
+  }
+
+  return this.lift(new ReduceOperator(accumulator, seed, hasSeed));
 }
 
 export class ReduceOperator<T, R> implements Operator<T, R> {
-
-  constructor(private accumulator: (acc: R, value: T) => R, private seed?: R) {
-  }
+  constructor(private accumulator: (acc: R, value: T) => R, private seed?: R, private hasSeed: boolean = false) {}
 
   call(subscriber: Subscriber<R>, source: any): any {
-    return source._subscribe(new ReduceSubscriber(subscriber, this.accumulator, this.seed));
+    return source._subscribe(new ReduceSubscriber(subscriber, this.accumulator, this.seed, this.hasSeed));
```

Comment:
we would have to have additional branching, would we not?

```js
if (this.hasSeed) {
  return source._subscribe(new ReduceSubscriber(subscriber, this.accumulator, this.seed));
} else {
  return source._subscribe(new ReduceSubscriber(subscriber, this.accumulator));
}
```

And then each class downstream would have to duplicate the `arguments.length >= 2` check wouldn't they?

I did it this way because it seemed to me the best performant and the other solutions had a lot of duplication. It's very non-obvious at first, but best I can tell you _have_ to check at the original call site otherwise you'll lose that info because of the `Operator#call` architecture.